src/datumaro/plugins/configurable_validator.py

```diff
@@ -0,0 +1,967 @@
+        self.imbalance_ratio_thr = (
+            imbalance_ratio_thr if imbalance_ratio_thr else self.DEFAULT_IMBALANCE_RATIO_THR
+        )
+        self.far_from_mean_thr = (
+            far_from_mean_thr if far_from_mean_thr else self.DEFAULT_FAR_FROM_MEAN_THR
+        )
+        self.dominance_thr = (
+            dominance_ratio_thr if dominance_ratio_thr else self.DEFAULT_DOMINANCE_RATIO_THR
+        )
+        self.topk_bins_ratio = topk_bins if topk_bins else self.DEFAULT_TOPK_BINS
+
+    def _init_stats_collector(self, label_categories):
+        for task in self.tasks:
+            kwargs = {
+                "few_samples_thr": self.few_samples_thr,
+                "imbalance_ratio_thr": self.imbalance_ratio_thr,
+                "far_from_mean_thr": self.far_from_mean_thr,
+                "dominance_thr": self.dominance_thr,
+                "topk_bins_ratio": self.topk_bins_ratio,
+            }
+            if task == TaskType.classification:
+                self.all_stats[task] = ClsStats(
+                    label_categories=label_categories, warnings=self.warnings, **kwargs
+                )
+            elif task == TaskType.detection:
+                self.all_stats[task] = DetStats(
+                    label_categories=label_categories, warnings=self.warnings, **kwargs
+                )
+            elif task == TaskType.segmentation:
+                self.all_stats[task] = SegStats(
+                    label_categories=label_categories, warnings=self.warnings, **kwargs
+                )
+        self.init_flag = True
+
+    def _get_stats_collector(self, ann_type):
+        if TaskType.classification in self.tasks and ann_type == AnnotationType.label:
+            return self.all_stats[TaskType.classification]
+        elif TaskType.detection in self.tasks and ann_type == AnnotationType.bbox:
+            return self.all_stats[TaskType.detection]
+        elif TaskType.segmentation in self.tasks and ann_type in [
+            AnnotationType.mask,
+            AnnotationType.polygon,
+            AnnotationType.ellipse,
+        ]:
+            return self.all_stats[TaskType.segmentation]
+        else:
+            return None
+
+    def compute_statistics(self, dataset):
+        if not self.init_flag:
+            self._init_stats_collector(label_categories=dataset.categories()[AnnotationType.label])
+
+        for item in dataset:
+            for stats_collector in self.all_stats.values():
+                if not stats_collector:
+                    continue
+                stats_collector.update_item(item)
+
+            item_key = (item.id, item.subset)
+            for annotation in item.annotations:
+                stats_collector = self._get_stats_collector(ann_type=annotation.type)
+                if not stats_collector:
+                    continue
+                stats_collector.update_ann(item_key, annotation)
```

Comment:
I haven't caught it up before, but it is very weird why there should exist the two paths between item update and annotation update at the top level although both can take `item` as their inputs. However, it requires many works to be fixed here, so that let's revisit afterwards.

Reply:
I believe, by introducing update_ann separately, we can reduce the annotation iterator. If we use only update_item, we have to run annotation iterator several times according to the desired task types. What do you think?

Reply:
Yes, it is fair point since Python has severe overheads on calling iterations.